mesh/nurbs.cpp

```diff
@@ -341,6 +341,100 @@ void KnotVector::CalcDnShape(Vector &gradn, int n, int i, double xi) const
 
 }
 
+void KnotVector::FindMaxima(Array<int> &ks,
+                            Vector &xi,
+                            Vector &u)
+{
+   Vector shape(Order+1);
+   Vector maxima(GetNCP());
+   double arg1,arg2,arg,max1,max2,max;
+
+   xi.SetSize(GetNCP());
+   u.SetSize(GetNCP());
+   ks.SetSize(GetNCP());
+   for (int j = 0; j <GetNCP(); j++)
+   {
+      maxima[j] = 0;
+      for (int d = 0; d < Order+1; d++)
+      {
+         int i = j - d;
+         if (isElement(i))
+         {
+            arg1 = 0;
+            CalcShape ( shape, i, arg1);
+            max1 = shape[d];
+
+            arg2 = 1;
+            CalcShape ( shape, i, arg2);
+            max2 = shape[d];
+
+            arg = (arg1 + arg2)/2;
+            CalcShape ( shape, i, arg);
+            max = shape[d];
+
+            while ( ( max > max1 ) || (max > max2) )
+            {
+               if (max1 < max2)
+               {
+                  max1 = max;
+                  arg1 = arg;
+               }
+               else
+               {
+                  max2 = max;
+                  arg2 = arg;
+               }
+
+               arg = (arg1 + arg2)/2;
+               CalcShape ( shape, i, arg);
+               max = shape[d];
+            }
+
+            if (max > maxima[j])
+            {
+               maxima[j] = max;
+               ks[j] = i;
+               xi[j] = arg;
+               u[j]  = getKnotLocation(arg, i+Order);
+            }
+         }
+      }
+   }
+}
+
+// Routine from "The NURBS book" - 2nd ed - Piegl and Tiller
```

Comment:
Finding a particular routine in the NURBS book can be time-consuming. Could you please add more details on where to find it? For example, these changes would also be helpful:
+// Routine from "The NURBS book" - 2nd ed - Piegl and Tiller (Algorithm A2.2)
 void KnotVector::CalcShape(Vector &shape, int i, double xi) const

+// Routine from "The NURBS book" - 2nd ed - Piegl and Tiller (Algorithm A2.3)
 void KnotVector::CalcDShape(Vector &grad, int i, double xi) const

Reply:
Source of added algorithms added in [59fbbe0](https://github.com/mfem/mfem/pull/3232/commits/59fbbe08fa936fc269f136e6cd1342ae6827113d)

Reply:
Other suggestions added as well: [0313c71](https://github.com/mfem/mfem/pull/3232/commits/0313c718d871bc4b3b8803db162d3de018859843)